Gibbs Ensemble scripts

jhossbach opened the pull request:
This is an updated version of #3903 

Several updates were performed:
- Multiprocessing and Communication via the internal Pipes was used instead of subprocessing and communicating via socket
- The sample system no longer uses correction terms but a truncation of 5 sigma instead

This updated way of communication between processes is far more efficient than using TCP connections which is either a result of an increase in latency or the multiprocessing module uses internal methods to efficiently use nonblocking communication between the processes.

The simulation has the following structure:
 - The run_sim.py script starts the clients and controls them (same functionality as the socket script in #3903)
 - The gibbs_client_class.py script contains the client class with the methods needed
 - The client_system.py script initializes the system in every box

jonaslandsgesell commented:
Would you mind investigating using MD moves instead of MC moves to propagate the internal states of each system. Just instead of moving 1 particle and computing exp(-beta Delta E_pot) in an MC step, run x MD steps (e.g. 10-100) and compute exp(-beta Delta E_pot) from before and after running MD. 
This would allow to simulate polymer systems and should have much faster decaying autocorrelations of observables.

Running MD will require that you initialize to the velocities with a Maxwellian distribution like e.g. in https://github.com/espressomd/espresso/blob/ae3094f47cf78cfafaddb5cdc8c0e7edb5630b4f/src/core/reaction_methods/ReactionAlgorithm.cpp#L503 after creation or when you insert new particles etc.

You will also have to avoid inserting particles closer than e.g. 0.9 sigma to avoid instabilities in the MD (see  https://github.com/espressomd/espresso/blob/ae3094f47cf78cfafaddb5cdc8c0e7edb5630b4f/src/core/reaction_methods/ReactionAlgorithm.cpp#L591 ) for fulfilling detailed balance you may then also not remove particles which were closer than this exclusion radius. You can just use the logic from within the reaction ensemble here.

RudolfWeeber commented:
I agree that MD moves will likely accelerate things. But let’s keep this Pr to the basics. Once it is in, we can extend upon it.

RudolfWeeber commented on samples/gibbs_ensemble/condor_queue.sub:
```diff
@@ -0,0 +1,59 @@
+Executable = run_for_several_kT.sh
```
I guess this is ICP specific. Please provide a simple bash script to run a few simulations for the rectilinear fit.

RudolfWeeber commented on samples/gibbs_ensemble/create_fits.py:
```diff
@@ -0,0 +1,105 @@
+import numpy as np
```
Please add file level docstrings to this and other files, explaining what is in that file. Please also add a Readme.md for the directory explaining what is ehere.

RudolfWeeber commented on samples/gibbs_ensemble/create_fits.py:
```diff
@@ -0,0 +1,105 @@
+# Warmup length to skip at beginning
+WARMUP_LENGTH = int(3e5)
+
+x = []
```
Please add comments on what the lists are for, or rename.

RudolfWeeber commented on samples/gibbs_ensemble/create_fits.py:
```diff
@@ -0,0 +1,105 @@
+kTs = []
+
+# 3D critical exponent
+beta = 0.32
```
Please mention the source (paper and equation nr) of the exponent and eqn.

RudolfWeeber commented on samples/gibbs_ensemble/create_fits.py:
```diff
@@ -0,0 +1,105 @@
+_dens_gas = []
+
+# Average over simulation results with different seeds
+for t in range(len(kTs)):
```
Please rewrite this in a more pythonic fashion:
* sort the lists by kT before the averaging
* loop over kTs directly in outer loop `for kt in kTs:`
* Use a list comprehensions to get densities for that kT: `[dens_liquid[i] for (i,t) in enumerate(kTs) if t == kT]` ro the like

jhossbach replied:
Should be resolved in commit 5e7f4e9, it is now as nested as goes at the cost of readabilty.

RudolfWeeber commented on samples/gibbs_ensemble/create_fits.py:
```diff
@@ -0,0 +1,105 @@
+print("B, T_c: {}".format(fit_scaling))
+print("Fit uncertainty: {}".format(np.diag(p_cov_scaling)))
+
+#kT_c = fit_scaling[1]
```
Please remove commented out code, or add an explanation if it should stay as an alternative

RudolfWeeber commented on samples/gibbs_ensemble/gibbs_client_class.py:
```diff
@@ -0,0 +1,134 @@
+from enum import unique, Enum
```
The file contains two classes. I'd suggest renaming it to `gibbs_ensemble.py`

RudolfWeeber commented on samples/gibbs_ensemble/gibbs_client_class.py:
```diff
@@ -0,0 +1,134 @@
+
+
+@unique
+class MessageID(Enum):
```
Please add a docstring for the class

RudolfWeeber commented on samples/gibbs_ensemble/gibbs_client_class.py:
```diff
@@ -0,0 +1,134 @@
+        """ Revert last particle add """
+        self.system.part[self.system.part.highest_particle_id].remove()
+
+    def remove_particle(self):
```
Please use the `ParticleHandle.to_dict()` function to store all properties of the particle, not just position. Use ParticleHandle.update() in the revert function. This will make the implementation electrostatics-proof.

jhossbach replied:
Implemented in a613cb2

RudolfWeeber commented on samples/gibbs_ensemble/run_sim.py:
```diff
@@ -0,0 +1,461 @@
+    level=logging.DEBUG if args.debug else logging.INFO
+)
+
+# Number of moves, number of accepted moves
```
Please make the tried/accepted move counters dictionaires with the elements of the message type enum as keys. This way, you do not have to remember which index in the list corresponds to which move type.

jhossbach replied:
Resolved in 0140287

RudolfWeeber commented on samples/gibbs_ensemble/run_sim.py:
```diff
@@ -0,0 +1,461 @@
+
+    # Debug
+    logging.debug(
+        "Perform volume change: Box {}: {} -> {}; Box {}: {} -> {}.".format(
```
Please mention paper/eqn for the move generation and acceptance probability

RudolfWeeber commented on samples/gibbs_ensemble/run_sim.py:
```diff
@@ -0,0 +1,461 @@
+    logging.debug("Particle move accepted.")
+
+    # Only if rand < acc
+    global num_accepted_moves
```
The use of `global` is discouraged. Please let the perform_...-functions return True or False depending on whether or not the move was accepted. You can update the counters outside the perform_...-functions then.

jhossbach replied:
Should be fixed in 0140287

RudolfWeeber commented on samples/gibbs_ensemble/run_sim.py:
```diff
@@ -0,0 +1,461 @@
+    num_accepted_moves[1] += 1
+
+
+def perform_particle_exchange(boxes):
```
Please mention paper/eqn which you implemented here.

RudolfWeeber commented on samples/gibbs_ensemble/run_sim.py:
```diff
@@ -0,0 +1,461 @@
+        - 1. / kT * recipient_box.diff_energy)
+
+    # Draw random number
+    rand = np.random.random()
```
Please outsource the move choosing into a function which returns one of the elements of the message enum. You can then use the result of the function also for the counter update.

jhossbach replied:
Please look at d441d87 for that

RudolfWeeber commented on samples/gibbs_ensemble/client_system.py:
```diff
@@ -0,0 +1,31 @@
+""" This script will set up the system in each box, use this to initialize interactions """
```
Instead of using a free funciton for setup, please inherit from the gibbs client class. In this way, you could (if needed) also modify other aspects of the client class, e.g., add a tail correction to the energy() function.

jhossbach replied:
Solved in 4291b46

jhossbach replied:
Solved in 4291b46

jhossbach commented:
For now I updated some comments, added docstrings and copyright disclaimers. Things that still need to be done:

- [x] Create README.md
- [x] Use subclass instead of free function in client_system to give rise to possible truncation correction
- [x] Actually implement truncation correction for better evaluation with the literature available
- [x] Rewrite the create_fits script to a more pythonic way
- [x] Add uncertainties to the plot
- [x] Add more commenting for lists throughout all scripts
- [x] Change the move counters to be Enum dict entries
- [x] Update the trial move functions to return True/False and no longer use the 'global' keyword
- [x] Update the Client class to save particle information via the ParticleHandle.to_dict() function

RudolfWeeber commented on samples/gibbs_ensemble/gibbs_ensemble.py:
```diff
@@ -0,0 +1,154 @@
+
+    def add_particle(self):
+        """ Add a particle at random inside box """
+        self.system.part.add(pos=np.random.random(3) * self.system.box_l)
```
Please store the handle to the created particle in the class and use it in the revert step. Otherwise, you might remove a different particle than what you added in case of a revert

jhossbach replied:
I like using handles to store the particle data, but why not use it for the other methods as well? This way one doesn't have to store the data with the to_dict() method

jhossbach replied:
I included what I meant in 2377506, please review it again.

RudolfWeeber commented on samples/gibbs_ensemble/gibbs_ensemble.py:
```diff
@@ -0,0 +1,154 @@
+
+    def revert_add_particle(self):
+        """ Revert last particle add """
+        self.system.part[self.system.part.highest_particle_id].remove()
```
self.last_added_particle.remove()
self.last_added_particle= None

RudolfWeeber commented on samples/gibbs_ensemble/client_system.py:
```diff
@@ -0,0 +1,79 @@
+    system.time_step = 0.01
+
+    for _ in range(num_part):
+        system.part.add(pos=np.random.random(3) * box_length)
```
Please use a single call to system.part.add() with an (N,3)-shaped array of random numbers for positions. Avoids the loop.

RudolfWeeber reviewed (dismissed):
Thank you. Just two minor issues left

jngrad commented on samples/gibbs_ensemble/run_for_several_kT.sh:
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
```
```suggestion
#!/bin/sh
```
For portability reasons, we prefer Dash over Bash whenever possible. In this case, the script can be ported to Dash by simply changing the conditional at line 6.

jngrad commented on samples/gibbs_ensemble/run_for_several_kT.sh:
```diff
@@ -0,0 +1,25 @@
+python=$(which python3)
+pypresso=../../build/pypresso
+
+if [[ ! -e $pypresso ]]
```
```suggestion
if [ ! -e "${pypresso}" ]
```
This change makes the script Dash-compatible. The quotation marks are required in Dash because the pypresso path may contain whitespaces (especially on Windows systems).

jngrad commented on samples/gibbs_ensemble/run_for_several_kT.sh:
```diff
@@ -0,0 +1,25 @@
+    for seed in 1 2 3 4 5
+    do
+
+        test -s "temp_$kT_$seed.dat.gz" ||
+            $pypresso run_sim.py $kT --log -steps 2000000 -seed $seed 2>&1 > "temp_$kT_seed_$seed.out" &
```
On Bash 5.0.17, this code doesn't work, because `"temp_$kT_$seed.dat.gz"` expands variable `$kT_` to an empty string, when in reality we want to expand variable `$kT`. Please consider using the tool [koalaman/shellcheck](https://github.com/koalaman/shellcheck), it helped us a lot fixing buggy shell scripts in the past. Also, the pypresso variable needs to be in quotes in case it contains whitespaces, and the .dat.gz file had the wrong filename.
```suggestion
        test -s "temp_${kT}_seed_${seed}.dat.gz" ||
            "${pypresso}" run_sim.py ${kT} --log -steps 2000000 -seed ${seed} 2>&1 > "temp_${kT}_seed_${seed}.out" &
```

jhossbach replied:
Thank you for the shellcheck tool, I will keep it in mind for future scripts.

jngrad commented on samples/gibbs_ensemble/run_sim.py:
```diff
@@ -0,0 +1,517 @@
+    default=0.3,
+    help="Init density, default 0.3")
+parser.add_argument(
+    '-steps',
+    '-s',
```
Here and elsewhere, it is quite common to have two dashes for the long name and a single dash for the abbreviation, i.e. `--steps / -s`.

jngrad replied:
...but then you also need to adapt `samples/gibbs_ensemble/run_for_several_kT.sh` at line 19 :-)

jngrad commented on samples/gibbs_ensemble/run_sim.py:
```diff
@@ -0,0 +1,517 @@
+#
```
These new files need to be tested in CI. The file to update is `testsuite/scripts/samples/test_gibbs_ensemble.py`.

jhossbach replied:
The `testsuite/scripts/samples/test_gibbs_ensemble.py` was looking for the old gibbs ensemble scripts, I updated the code to run the right code in 6ca1ba0.

jngrad commented on samples/gibbs_ensemble/run_for_several_kT.sh:
```diff
@@ -0,0 +1,25 @@
+wait
+
+# Plot the results, create fits
+$python create_fits.py temp_*.dat.gz
```
```suggestion
"${python}" create_fits.py temp_*.dat.gz
```

jngrad reviewed (commented):
Here is a review from the point of view of code maintainability. I'm not sure the shell script can be tested, unless we can extract the number of steps as an external variable and call the script from a python subprocess.

jngrad commented:
@jhossbach please apply suggested changes with option ["Add suggestion to batch"](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/incorporating-feedback-in-your-pull-request#applying-suggested-changes), we have limited CI resources :-)

RudolfWeeber commented:
>I like using handles to store the particle data, but why not use it for the other methods as well? This way one doesn't have to store the data with the to_dict() >method

The handle is a proxy to the actual particle in the Espresso system. Once you delete the particle, the handle is meaningless. Hence, just storing the handle is not enough information to revert a removal of a particle. You have to store the actual particle properties (as a copy), before removing the particle.

jhossbach commented:
With fe1f069 I did a last sweep over the plot script. Please review it again.

> Thank you. Just two minor issues left

KaiSzuttor commented on samples/gibbs_ensemble/client_system.py:
```diff
@@ -0,0 +1,87 @@
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    @property
+    def energy(self):
+        """ Additional correction terms """
+        return super().energy  # + \
+        # calc_tail_correction(self.density) + \
+        # calc_shift_correction(self.density)
```
is it intended to comment out the corrections? if you are not changing anything in the inherited method you should just omit it

jhossbach replied:
The commented out version serves a better comparison with the literature as often the correction terms are omitted, for transparency i planned to define it but comment it out. Let me know what you think.

jngrad replied:
You can look at it this way: how can we know if the functions `calc_tail_correction()` and `calc_shift_correction()` do what they claim to be doing if they are not executed in the tests? We regularly do maintenance operations on samples, e.g. small refactoring projects, code simplification and code formatting. Unfortunately these operations always introduce bugs, and we rely on the testsuite to catch these mistakes.

jhossbach replied:
Fair enough, I will delete the functions in my next commit as the corrections will not be needed for the sample code.

jngrad replied:
Removed in 1bad96b8fa66a00845e.

KaiSzuttor commented on samples/gibbs_ensemble/create_fits.py:
```diff
@@ -0,0 +1,197 @@
+
+# Scaling law and Rectilinear law as shown in Frenkel, Smit, eqs. (8.3.6)
+# and (8.3.7)
+def scaling_law_fit(kT, B, kT_c):
```
please move the comment into the (missing) docstring of the function
```suggestion
def scaling_law(kT, B, kT_c):
```

KaiSzuttor commented on samples/gibbs_ensemble/create_fits.py:
```diff
@@ -0,0 +1,197 @@
+
+
+def exp(x, a):
+    return np.exp(-a * x)
```
either use a lambda for such a simple expression or move the function into ```get_relaxation_time```

KaiSzuttor commented on samples/gibbs_ensemble/create_fits.py:
```diff
@@ -0,0 +1,197 @@
+    return np.exp(-a * x)
+
+
+def get_relaxation_time(x):
```
```suggestion
def relaxation_time(x):
```

KaiSzuttor commented on samples/gibbs_ensemble/create_fits.py:
```diff
@@ -0,0 +1,197 @@
+    return popt[0]
+
+
+def compute_error(x):
```
```suggestion
def std_error_of_mean(x):
```

KaiSzuttor replied:
please also add a docstring

KaiSzuttor commented on samples/gibbs_ensemble/create_fits.py:
```diff
@@ -0,0 +1,197 @@
+
+# 3D critical exponent (See Frenkel, Smits: "Understanding Molecular
+# simulation", p.217
+beta = 0.32
```
if this is a constant
```suggestion
BETA = 0.32
```

KaiSzuttor commented on samples/gibbs_ensemble/create_fits.py:
```diff
@@ -0,0 +1,197 @@
+    return B * np.abs(kT - kT_c)**beta
+
+
+def rectilinear_law_fit(kT, p_c, A, kT_c):
```
docstring

KaiSzuttor commented on samples/gibbs_ensemble/create_fits.py:
```diff
@@ -0,0 +1,197 @@
+# Load data
+for f in args.files:
+
+    print("Loading {}...".format(f))
```
consider using the ```logging``` feature of python

KaiSzuttor commented on samples/gibbs_ensemble/create_fits.py:
```diff
@@ -0,0 +1,197 @@
+for f in args.files:
+
+    print("Loading {}...".format(f))
+    with gzip.open(f, 'rb') as _f:
```
consider using a try except block here and throw a readable error if the data is not there

KaiSzuttor commented on samples/gibbs_ensemble/create_fits.py:
```diff
@@ -0,0 +1,197 @@
+
+
+# Load data
+for f in args.files:
```
instead of global variables and module level statements consider using functions and return the results

jhossbach replied:
I find it more clear to extract and store all data from the files in one block rather than outsourcing it to functions, please leave a suggestion how this could be handled better.

RudolfWeeber replied:
In a larger software the single file extraction should definitely be in a function. For a plotting script, I'd say this is OK.

KaiSzuttor commented on samples/gibbs_ensemble/gibbs_ensemble.py:
```diff
@@ -0,0 +1,157 @@
+@unique
+class MessageID(Enum):
+    # Message identifiers
+    ENERGY = 0
+    END = 1
+    MOVE_PART = 2
+    MOVE_PART_REVERT = 21
+    CHANGE_VOLUME = 3
+    CHANGE_VOLUME_REVERT = 31
+    PART_ADD = 4
+    PART_ADD_REVERT = 41
+    PART_REMOVE = 5
+    PART_REMOVE_REVERT = 51
+    INFO = 6
```
do these enum values have a special meaning? if not consider using automatic generation of the values via ```enum.auto()```

KaiSzuttor reviewed (commented):
looks very good, i just have added a few minor comments

RudolfWeeber commented:
From my side, we're good, here.

jngrad commented:
I'll have another look and rebase the PR. There are 68 commits, many of which are fixups and reverts.

jhossbach committed:
samples: New Gibbs Ensemble scripts

jngrad committed:
samples: Reduce Gibbs Ensembe code complexity

jngrad commented on samples/gibbs_ensemble/run_sim.py:
```diff
@@ -0,0 +1,524 @@
+try:
+    assert((GLOBAL_NUM_PART % 2) == 0)
+except AssertionError:
+    raise TypeError("Wrong number of particles, must be even")
```
The error message can be written in the assertion itself. The syntax is `assert EXPRESSION, MESSAGE`. Note there are no round braces, because `assert` is not a function. Adding round braces can lead to nasty bugs, here is an illustration:
```python
def foo(run_dangerous_function):
   assert run_dangerous_function, "You shouldn't run a dangerous function"
   drop_table()

def bar(run_dangerous_function):
   assert(run_dangerous_function, "You shouldn't run a dangerous function")
   drop_table()

foo(False) # safe: the assertion caught the False
bar(False) # oops!
```

jngrad commented on samples/gibbs_ensemble/run_sim.py:
```diff
@@ -0,0 +1,524 @@
+        except BaseException:
+            raise ConnectionError(
+                "Error during energy return of {}, got this instead: \n{}".format(
+                    self.box_name, msg))
```
simply use an if statement

jngrad commented on samples/gibbs_ensemble/create_fits.py:
```diff
@@ -0,0 +1,210 @@
+        with gzip.open(f, 'rb') as _f:
+            data = pickle.load(_f)
+    except FileNotFoundError:
+        raise FileNotFoundError("File {} not found.".format(f))
```
Redundant, the original python `FileNotFoundError` already prints the filename.

jngrad reviewed (commented):
A few suggestions. They have been addressed in 684a12d830e9ba8b20758d993a4c8ca1ef2153d4

jngrad commented:
The PR was rebased to remove accidentally committed files from the git history and squash the fixups. I've also added copyright headers and wrapped long lines (see `git diff 02c06cbc94c..7fe8f511489233`).

jngrad committed:
samples: Remove unused functions

jngrad reviewed (approved):
ok from my side

jngrad commented:
@RudolfWeeber don't forget to change your review from "Requested changes" to "Accepted", then it can be merged

kodiakhq[bot] committed:
Merge branch 'python' into python
